flow/include/flow/FlowThread.h

```diff
@@ -0,0 +1,294 @@
+/*
+ * FlowThread.h
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2025 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FLOW_FLOW_THREAD_H
+#define FLOW_FLOW_THREAD_H
+#include "flow/Buggify.h"
+#pragma once
+
+#include <flow/flow.h>
+#include <flow/FastAlloc.h>
+#include <flow/ThreadPrimitives.h>
+#include <flow/ThreadHelper.actor.h>
+
+// NOTE: Currently futures should only be used from main thread.
+template <class T>
+class ThreadNotifiedQueue : private SingleCallback<T>, public FastAllocated<ThreadNotifiedQueue<T>> {
+public:
+	int promises;
+	int futures;
+	Error error;
+
+	// Invariant: SingleCallback<T>::next==this || (queue.empty() && !error.isValid())
+	std::queue<T, Deque<T>> queue;
+	ThreadSpinLock mutex;
+
+	ThreadNotifiedQueue(int futures, int promises) : promises(promises), futures(futures) {
+		SingleCallback<T>::next = this;
+	}
+
+	virtual ~ThreadNotifiedQueue() = default;
+
+	bool isReady() {
+		ThreadSpinLockHolder holder(mutex);
+		return !queue.empty() || error.isValid();
+	}
+
+	bool isError() {
+		ThreadSpinLockHolder holder(mutex);
+		return queue.empty() && error.isValid();
+	} // the *next* thing queued is an error
```

Comment:
is an error or 'gets' an error?

Reply:
Didn't quite understand the question. It is basically doing what original implementation was doing. I think the idea is to let the runtime empty the queue before it realizes there is error. So for example, you put 10 values, and next one end's up being some sort of error, this is used by runtime so that it pops the value that were sent before the error was sent.

Reply:
I was asking about the comment. I don't understand what it is trying to say. We are putting an error in the queue "the *next* thing queued is an error")? Or is it any attempt to add to the queue will now return in error?